### PR TITLE
Pin a specific python version on Github Actions

### LINKED ISSUE #9281: Python version linked to on Github Actions is inconsistent
Issue overview
--------------

While we do use actions/setup-python@v2, cmake is likely going to find another version.

### Details
Some additional details for this issue (if relevant):
 - Platform (Operating system, version)
 - Version of EnergyPlus (if using an intermediate build, include SHA)
 - Unmethours link or helpdesk ticket number

### Checklist
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [ ] Defect file added (list location of defect file here)
 - [x] Ticket added to Pivotal for defect (development team task)
 - [x] Pull request created (the pull request will have additional tasks related to reviewing changes that fix this defect)

### PR DESCRIPTION
**Pull request by jmarrec**
Pull request overview
---------------------

* Fix #9281

Pin a specific python version on Github Actions. Currently at 3.8. This ensures that whatever python version we request from actions/setup-python@v2 (there is an env variable atop the workflow file) is actually what cmake picks up.

Mac is the one that gave me trouble, because the system python is also 3.8. But I got it to comply by actually passing both the python version and the Python_ROOT_DIR (as a hint to FindPython)

See these tests on my fork:

- windows x86: https://github.com/jmarrec/EnergyPlus/runs/5238587515?check_suite_focus=true#step:9:117
- windows x64: https://github.com/jmarrec/EnergyPlus/runs/5238587466?check_suite_focus=true#step:9:117
- mac 10.15 https://github.com/jmarrec/EnergyPlus/runs/5238587224?check_suite_focus=true#step:8:113
- mac 11.6 https://github.com/jmarrec/EnergyPlus/runs/5238587272?check_suite_focus=true#step:8:113
- Ubuntu 18.04: https://github.com/jmarrec/EnergyPlus/runs/5238587220?check_suite_focus=true#step:8:103
- Ubuntu 20.04: https://github.com/jmarrec/EnergyPlus/runs/5238587279?check_suite_focus=true#step:8:103

| Runner      | python | Python Path found by cmake                                    |
|-------------|--------|---------------------------------------------------------------|
| mac         | 3.8.12 | /Users/runner/hostedtoolcache/Python/3.8.12/x64/bin/python3.8 |
| linux       | 3.8.12 | /opt/hostedtoolcache/Python/3.8.12/x64/bin/python3.8          |
| windows x64 | 3.8.10 | C:/hostedtoolcache/windows/Python/3.8.10/x64/python3.exe      |
| windows x86 | 3.8.10 | C:/hostedtoolcache/windows/Python/3.8.10/x84/python3.exe      |

### Pull Request Author
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [ ] Title of PR should be user-synopsis style (clearly understandable in a standalone changelog context)
 - [ ] Label the PR with at least one of: Defect, Refactoring, NewFeature, Performance, and/or DoNoPublish
 - [ ] Pull requests that impact EnergyPlus code must also include unit tests to cover enhancement or defect repair
 - [ ] Author should provide a "walkthrough" of relevant code changes using a GitHub code review comment process
 - [ ] If any diffs are expected, author must demonstrate they are justified using plots and descriptions
 - [ ] If changes fix a defect, the fix should be demonstrated in plots and descriptions
 - [ ] If any defect files are updated to a more recent version, upload new versions here or on DevSupport
 - [ ] If IDD requires transition, transition source, rules, ExpandObjects, and IDFs must be updated, and add IDDChange label
 - [ ] If structural output changes, add to output rules file and add OutputChange label
 - [ ] If adding/removing any LaTeX docs or figures, update that document's CMakeLists file dependencies

### Reviewer
This will not be exhaustively relevant to every PR.
 - [ ] Perform a Code Review on GitHub
 - [ ] If branch is behind develop, merge develop and build locally to check for side effects of the merge
 - [ ] If defect, verify by running develop branch and reproducing defect, then running PR and reproducing fix
 - [ ] If feature, test running new feature, try creative ways to break it
 - [ ] CI status: all green or justified
 - [ ] Check that performance is not impacted (CI Linux results include performance check)
 - [ ] Run Unit Test(s) locally
 - [ ] Check any new function arguments for performance impacts
 - [ ] Verify IDF naming conventions and styles, memos and notes and defaults
 - [ ] If new idf included, locally check the err file and other outputs

### CODE REVIEW

File: .github/workflows/linux_release.yml [hunk 1]
```diff
@@ -8,6 +8,7 @@ on:
 env:
   GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
   BUILD_TYPE: Release
+  PYTHON_VERSION: 3.8
```
Comment (jmarrec) [inline .github/workflows/linux_release.yml:11]: New env version for all runners, so there's a single place to modify it

File: .github/workflows/linux_release.yml [hunk 2]
```diff
@@ -19,10 +20,11 @@ jobs:
     steps:
     - uses: actions/checkout@v2
 
-    - name: Set up Python 3.7
+    - name: Set up Python ${{ env.PYTHON_VERSION }}
+      id: setup-python
       uses: actions/setup-python@v2
       with:
-        python-version: 3.7
+        python-version: ${{ env.PYTHON_VERSION }}
```
Comment (jmarrec) [inline .github/workflows/linux_release.yml:27]: Use the env version to setup python

File: .github/workflows/linux_release.yml [hunk 3]
```diff
@@ -49,9 +51,11 @@ jobs:
       working-directory: ./build
       shell: bash
       run: |
-        cmake -DCMAKE_BUILD_TYPE=$BUILD_TYPE \
-              -DLINK_WITH_PYTHON=ON -DBUILD_FORTRAN=ON -DBUILD_PACKAGE:BOOL=ON \
-              -DDOCUMENTATION_BUILD="BuildWithAll" -DTEX_INTERACTION="batchmode" -DENABLE_PCH=OFF \
+        cmake -DCMAKE_BUILD_TYPE:STRING=$BUILD_TYPE \
+              -DLINK_WITH_PYTHON:BOOL=ON -DPYTHON_VERSION:STRING=${{ steps.setup-python.outputs.python-version }} \
+              -DPython_ROOT_DIR=$RUNNER_TOOL_CACHE/Python/${{ steps.setup-python.outputs.python-version }}/x64/ \
```
Comment (jmarrec) [inline .github/workflows/linux_release.yml:56]: Pass the PYTHON_VERSION cmake variable to make cmake request an EXACT match. We request the **actual** version that was picked by actions/setup-python. So we request 3.8 for eg, and we get 3.8.12 from setup-python and use 3.8.12 here

File: CMakeLists.txt [hunk 4]
```diff
@@ -159,7 +159,13 @@ endif()
...
 # If LINK_WITH_PYTHON, also request the Development (libs) at the same time, to ensure consistent version between interpreter and Development
 if(LINK_WITH_PYTHON)
-  find_package(Python 3.6 COMPONENTS Interpreter Development REQUIRED)
+  # find_package(Python) has the problem that on github actions in particular it'll pick up the most recent python (eg 3.9) from the tool cache
+  # even if you have used the setup-python action and set it to 3.8, so we make the exact version required
+  if(PYTHON_VERSION)
+    find_package(Python ${PYTHON_VERSION} EXACT COMPONENTS Interpreter Development REQUIRED)
+  else()
+    find_package(Python 3.6 COMPONENTS Interpreter Development REQUIRED)
+  endif()
```
Comment (jmarrec) [inline CMakeLists.txt:168]: Undocumented `PYTHON_VERSION` cmake variable (do we want to have it show in ccmake/cmake gui instead?).

If passed, we request an EXACT version

**Comment by Myoldmopar** [comment]
Decent CI looked happy with the change.  Thanks for the fix to get the Linux GHA build going.  Assuming everything is happy I'll merge this when Decent is done with it this evening.

**Comment by Myoldmopar** [comment]
GHA alternate Linux build passed with flying colors.  Again, if Decent starts passing as well then this will drop later today.

**Comment by Myoldmopar** [comment]
All happy!  Merging this one, thanks @jmarrec !

**Pull request merged**